gdal/swig/python/osgeo/utils/gdal_calc.py

```diff
@@ -129,57 +204,139 @@ def doit(opts, args):
     # set up global namespace for eval with all functions of gdalnumeric
     global_namespace = dict([(key, getattr(gdalnumeric, key))
                              for key in dir(gdalnumeric) if not key.startswith('__')])
+    if opts.user_namespace:
+        global_namespace.update(opts.user_namespace)
 
     if not opts.calc:
         raise Exception("No calculation provided.")
     elif not opts.outF:
-        raise Exception("No output file provided.")
+        if not opts.return_ds:
+            raise Exception("No output file provided.")
+        else:
+            os.format = 'MEM'
 
     if opts.format is None:
         opts.format = GetOutputDriverFor(opts.outF)
 
+    if not hasattr(opts, "color_table"):
+        opts.color_table = None
+    if not opts.extent:
+        opts.extent = EXTENT_IGNORE
+    else:
+        opts.extent = parse_extent(opts.extent)
+
+    equal_gt_eps = 0.000001
+    gt_diff_support = (False, False, False, False) if opts.extent == EXTENT_FAIL else (True, False, False, False)
+    gt_diff_error = (
```

Comment:
I would rather make this is a dict, with explicit use of the GT_ constants to avoid relying on their value, which is fragile

Reply:
done